File: Older_DEW4SLHA_versions/src/DHS_calc.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Jun  8 14:33:53 2023

Delta_HS calculator.

@author: Dakotah Martinez
"""

import numpy as np

def Delta_HS_calc(mHdsq_Lambda, delta_mHdsq, mHusq_Lambda, delta_mHusq,
                  mu_Lambdasq, delta_musq, running_mz_sq, tanb_sq, sigmauutot,
                  sigmaddtot):
    """
    Compute the fine-tuning measure Delta_HS.

    Parameters
    ----------
    mHdsq_Lambda : Float.
        mHd^2(GUT).
    delta_mHdsq : Float.
        RGE running of mHd^2 down to 2 TeV.
    mHusq_Lambda : Float.
        mHu^2(GUT).
    delta_mHusq : Float.
        RGE running of mHu^2 down to 2 TeV.
    mu_Lambdasq : Float.
        mu^2(GUT).
    delta_musq : Float.
        RGE running of mu^2 down to 2 TeV.
    running_mz_sq : Float.
        Running mZ^2, evaluated at 2 TeV.
    tanb_sq : Float.
        tan^2(beta), evaluated at 2 TeV.
    sigmauutot : Float.
        Up-type radiative corrections evaluated at 2 TeV.
    sigmaddtot : Float.
        Down-type radiative corrections evaluated at 2 TeV.

    Returns
    -------
    Delta_HS : Float.
        Fine-tuning measure Delta_HS.

    """
    B_Hd = mHdsq_Lambda / (tanb_sq - 1)
    B_deltaHd = delta_mHdsq / (tanb_sq - 1)
    B_Hu = mHusq_Lambda * tanb_sq / (tanb_sq - 1)
    B_deltaHu = delta_mHusq * tanb_sq / (tanb_sq - 1)
    B_Sigmadd = sigmaddtot / (tanb_sq - 1)
    B_Sigmauu = sigmauutot * tanb_sq / (tanb_sq - 1)
    B_muLambdasq = mu_Lambdasq
    B_deltamusq = delta_musq
    Delta_HS_contribs = np.sort(np.array([(np.abs(B_Hd) / (running_mz_sq / 2),
                                           'Delta_HS(mHd^2(GUT))'),
                                          (np.abs(B_deltaHd)
                                           / (running_mz_sq / 2),
                                           'Delta_HS(delta(mHd^2))'),
                                          (np.abs(B_Hu)/ (running_mz_sq / 2),
                                           'Delta_HS(mHu^2(GUT))'),
                                          (np.abs(B_deltaHu)
                                           / (running_mz_sq / 2),
                                           'Delta_HS(delta(mHu^2))'),
                                          (np.abs(B_Sigmadd)
                                           / (running_mz_sq / 2),
                                           'Delta_HS(Sigma_d^d)'),
                                          (np.abs(B_Sigmauu)
                                           / (running_mz_sq / 2),
                                           'Delta_HS(Sigma_u^u)'),
                                          (np.abs(B_muLambdasq)
                                           / (running_mz_sq / 2),
                                           'Delta_HS(mu(GUT))'),
                                          (np.abs(B_deltamusq)
                                           / (running_mz_sq / 2),
                                           'Delta_HS(delta(mu))')],
                                         dtype=[('HSContrib', float),
                                                ('HSlabel', 'U40')]),
                                order='HSContrib')
    return Delta_HS_contribs[::-1]

File: natLHA/src/DHS_calc.cpp
#include <vector>
#include <string>
#include <algorithm>
#include <cmath>
#include "DHS_calc.hpp"
#include <boost/multiprecision/mpfr.hpp>
#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

using namespace boost::multiprecision;
typedef number<mpfr_float_backend<50>> high_prec_float;

bool absValCompareHS(const LabeledValueHS& a, const LabeledValueHS& b) {
    return abs(a.value) < abs(b.value);
}

std::vector<LabeledValueHS> sortAndReturnHS(const std::vector<LabeledValueHS>& DHSList) {
    std::vector<LabeledValueHS> sortedList = DHSList;
    sort(sortedList.begin(), sortedList.end(), absValCompareHS);
    reverse(sortedList.begin(), sortedList.end());
    return sortedList;
}

std::vector<LabeledValueHS> DHS_calc(high_prec_float& mHdsq_Lambda, high_prec_float delta_mHdsq, high_prec_float& mHusq_Lambda, high_prec_float delta_mHusq,
                                high_prec_float mu_Lambdasq, high_prec_float delta_musq, high_prec_float running_mZ_sq, high_prec_float tanb_sq, high_prec_float& sigmauu_tot,
                                high_prec_float& sigmadd_tot) {
    high_prec_float B_Hd = mHdsq_Lambda / (tanb_sq - 1.0);
    high_prec_float B_Hu = mHusq_Lambda * tanb_sq / (tanb_sq - 1.0);
    high_prec_float B_Sigmadd = sigmadd_tot / (tanb_sq - 1.0);
    high_prec_float B_Sigmauu = sigmauu_tot * tanb_sq / (tanb_sq - 1.0);
    high_prec_float B_muLambdasq = mu_Lambdasq;
    
    high_prec_float B_deltaHd = delta_mHdsq / (tanb_sq - 1.0);
    high_prec_float B_deltaHu = delta_mHusq * tanb_sq / (tanb_sq - 1.0);
    high_prec_float B_deltamusq = delta_musq;

    std::vector<LabeledValueHS> Delta_HS_contribs = {{B_Hd / (pow(high_prec_float(911876.0) / high_prec_float(10000.0), high_prec_float(2.0)) / high_prec_float(2.0)), "Delta_HS(mHd^2(GUT))"},
                                                {B_Hu / (pow(high_prec_float(911876.0) / high_prec_float(10000.0), high_prec_float(2.0)) / high_prec_float(2.0)), "Delta_HS(mHu^2(GUT))"},
                                                {B_muLambdasq / (pow(high_prec_float(911876.0) / high_prec_float(10000.0), high_prec_float(2.0)) / high_prec_float(2.0)), "Delta_HS(mu^2(GUT))"},
                                                {B_deltaHd / (pow(high_prec_float(911876.0) / high_prec_float(10000.0), high_prec_float(2.0)) / high_prec_float(2.0)), "Delta_HS(delta(mHd^2))"},
                                                {B_deltaHu / (pow(high_prec_float(911876.0) / high_prec_float(10000.0), high_prec_float(2.0)) / high_prec_float(2.0)), "Delta_HS(delta(mHu^2))"},
                                                {B_deltamusq / (pow(high_prec_float(911876.0) / high_prec_float(10000.0), high_prec_float(2.0)) / high_prec_float(2.0)), "Delta_HS(delta(mu^2))"},
                                                {B_Sigmadd / (pow(high_prec_float(911876.0) / high_prec_float(10000.0), high_prec_float(2.0)) / high_prec_float(2.0)), "Delta_HS(Sigma_d^d)"},
                                                {B_Sigmauu / (pow(high_prec_float(911876.0) / high_prec_float(10000.0), high_prec_float(2.0)) / high_prec_float(2.0)), "Delta_HS(Sigma_u^u)"}};
    std::vector<LabeledValueHS> sortedList = sortAndReturnHS(Delta_HS_contribs);
    return sortedList;
}